### PR TITLE
IHostingStartup topic

### LINKED ISSUE #4438: Document IHostStartup (host lightup feature)
See Add docs for IHostedService #3352

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #4438 

[Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/hosting/ihostingstartup?branch=pr-en-us-4908)

https://github.com/guardrex/HostingStartupSample

### CODE REVIEW

**Comment by guardrex** [comment]
@pakrym @Tratcher Let's TAL 👓 whenever ur ready and see how this is shaping up.

#### Notes:
* Not a huge fan of the title ... *and I wrote it!* :grinning: Idears?
* Uses a build target approach with a PS script to move files and mod the deps file. Script works on Win and Mac, but I can only confirm that the whole process works on Win. My 2009 Mac Mini can't take Sierra, so it can't take 2.0, so it can't really test this approach adequately for Mac. I did this to automate the sample as much as possible to avoid having to wrap the `IHostingStartup` project and trim the deps file. See if my approach is *sane* here. Vegas puts the odds at 50-50 that it's ok. :smile:
* I addressed the feedback in #4438.

File: aspnetcore/hosting/ihostingstartup.md [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+## Implement IHostingStartup features
+
+### Create the assembly
+
+An `IHostingStartup` feature is deployed as an assembly based on a console app without an entry point. The assembly references the [Microsoft.AspNetCore.Hosting.Abstractions](https://www.nuget.org/packages/Microsoft.AspNetCore.Hosting.Abstractions/) package:
+
+[!code-xml[Main](ihostingstartup/snapshot_sample/StartupFeature.csproj)]
+
+A [HostingStartupAttribute](/dotnet/api/microsoft.aspnetcore.hosting.hostingstartupattribute) identifies a class as an implementation of `IHostingStartup` for loading and execution when building the [IWebHost](/dotnet/api/microsoft.aspnetcore.hosting.iwebhost). In the following example, the namespace is `StartupFeature`, and the class is `StartupFeatureHostingStartup`:
```
Comment (Rick-Anderson) [inline aspnetcore/hosting/ihostingstartup.md]: `A [HostingStartup](/dotnet/api/microsoft.aspnetcore.hosting.hostingstartupattribute) attribute identifies a class as an implementation `

A [HostingStartup](/dotnet/api/microsoft.aspnetcore.hosting.hostingstartupattribute) attribute identifies a class as an implementation

We usually write attributes that way give that's how they're used in code.

File: aspnetcore/hosting/ihostingstartup.md [hunk 2]
```diff
@@ -85,7 +85,7 @@ For global use, place the file in the `additonalDeps` folder of the .NET Core in
 <DRIVE>\Program Files\dotnet\additionalDeps\<FEATURE_ASSEMBLY_NAME>\shared\Microsoft.NETCore.App\2.0.0\
 ```
 
-Note that the version, `2.0.0`, reflects the version of the shared runtime that the target app uses (the shared runtime listed in *\*.runtimeconfig.json*). In the sample app, the shared runtime is specified in the *HostingStartupSample.runtimeconfig.json* file.
+Note the version, `2.0.0`, reflects the version of the shared runtime that the target app uses. The shared runtime listed in *\*.runtimeconfig.json*. In the sample app, the shared runtime is specified in the *HostingStartupSample.runtimeconfig.json* file.
```
Comment (guardrex) [inline aspnetcore/hosting/ihostingstartup.md]: Missing "is"? ... perhaps with "file" and a def article:

> The shared runtime is shown in the \*\\\*.runtimeconfig.json\* file.

**Comment by guardrex** [comment]
> LGTM as long as a dev has signed off.

We need a real programmer to look at my *dino hacks* 🦎 on this one. :smile:

**Comment by guardrex** [comment]
@Rick-Anderson This wasn't signed-off by an engineering team member ... did they verbally say it was 'ok'?

**Comment by Tratcher** [comment]
Sorry I didn't get to this last week, I'm out of the office a lot this month.

The content here is primarily focused on implementers, it doesn't cover any of the consumer mechanics. We get a lot of questions from folks about how to control and debug this injection process and I think we should add that information here. See this thread https://github.com/aspnet/Home/issues/2051

**Comment by guardrex** [comment]
Moving discussion over to the new PR: #4963

**Pull request merged**